x-pack/plugins/security_solution/public/common/containers/sourcerer/index.tsx

```diff
@@ -72,29 +72,44 @@ export const useInitSourcerer = (
     getTimelineSelector(state, TimelineId.active)
   );
   const scopeIdSelector = useMemo(() => sourcererSelectors.scopeIdSelector(), []);
-  const { selectedDataViewId: scopeDataViewId } = useDeepEqualSelector((state) =>
-    scopeIdSelector(state, scopeId)
-  );
-  const { selectedDataViewId: timelineDataViewId } = useDeepEqualSelector((state) =>
-    scopeIdSelector(state, SourcererScopeName.timeline)
-  );
-  const activeDataViewIds = useMemo(
-    () => [...new Set([scopeDataViewId, timelineDataViewId])],
-    [scopeDataViewId, timelineDataViewId]
-  );
+  const {
+    selectedDataViewId: scopeDataViewId,
+    selectedPatterns,
+    missingPatterns,
+  } = useDeepEqualSelector((state) => scopeIdSelector(state, scopeId));
+  const {
+    selectedDataViewId: timelineDataViewId,
+    selectedPatterns: timelineSelectedPatterns,
+    missingPatterns: timelineMissingPatterns,
+  } = useDeepEqualSelector((state) => scopeIdSelector(state, SourcererScopeName.timeline));
   const { indexFieldsSearch } = useDataView();
 
   const searchedIds = useRef<string[]>([]);
-  useEffect(
-    () =>
-      activeDataViewIds.forEach((id) => {
-        if (id != null && id.length > 0 && !searchedIds.current.includes(id)) {
-          searchedIds.current = [...searchedIds.current, id];
-          indexFieldsSearch(id);
-        }
-      }),
-    [activeDataViewIds, indexFieldsSearch]
-  );
+  useEffect(() => {
+    const activeDataViewIds = [...new Set([scopeDataViewId, timelineDataViewId])];
+    activeDataViewIds.forEach((id) => {
+      if (id != null && id.length > 0 && !searchedIds.current.includes(id)) {
+        searchedIds.current = [...searchedIds.current, id];
+        indexFieldsSearch(
+          id,
+          id === scopeDataViewId ? SourcererScopeName.default : SourcererScopeName.timeline,
```

Comment:
![wut](https://media2.giphy.com/media/Efpa4tgdBl5DaQYVda/giphy.gif?cid=ecf05e47uxq9cf8ilk1pc9soqaovrxh7y983ua3kubnxwy9p&rid=giphy.gif&ct=g)

can you please add some comments with what you're doing here, and also move all of this into a useCallback and have the useEffect look like this:

```
useEffect(() => {
  onActiveDataViewIdsUpdate()
}, [activeDataViewIds]);
```

this way we are only calling all this when `activeDataViewIds` updates instead of all that long dependency list

Some jests tests around this would be great too! I already have `'index field search called twice when default and timeline have different dataViewId'`, but it would be good to get all these new conditions too

Reply:
I will add comment, but I won't do the callback because I need some of the dependency to be in the useEffect and we are protected by the history useRef to not re-run it, if we already did it.